### ISSUE TITLE
ADFS compatibility

### ISSUE THREAD

**Issue by maxh8086** [fydrah/loginapp#16]
Have tested Kubernetes with ADFS oAuth / OIDC Successfully, but to generate .kube/config via loginapp URL need addition to provide resource URL

**Comment by maxh8086** [comment]
![image](https://user-images.githubusercontent.com/48318337/53953546-9a396d80-40f9-11e9-8010-6f612a49e739.png)

![image](https://user-images.githubusercontent.com/48318337/53953565-a9b8b680-40f9-11e9-9e78-8c548ede2a0c.png)

Ref URL 
![image](https://user-images.githubusercontent.com/48318337/53953572-b210f180-40f9-11e9-88f1-a88b022e6f0a.png)

I personally have tested it with  &resource= using adfs to get email and group membership claims to make it working and has giving me right ID Token.

but doing it manually every time we request code is flaw and wish to automate it with the code itself.

looking forward for further help.

**Comment by fydrah** [comment]
Hi @maxh8086 ,

Based on your tests, could you give as example:
* What behaviour for loginapp do you expect
* What behaviour do you currently have

This will help to understand your problem.

Regards

**Comment by maxh8086** [comment]
We have configured Loginapp to point ADFS where in ADFS is sucessfully able to generate and provide us token, but the problem is without providing "Resource" in web Query (probably Get or Post) we do not get custom claims like email and group membership which fails to process further Authentication / Authorization with Kubernetes due to missing email / group claim.

ADFS Expect Resource in post where in resource can be Callback URL which user need to configure under ADFS or can be Kubernetes API URL which we by default provide while configuring application group.

Below is URL where we got this Reference and tried adding Resource manually before authentication.

![image](https://user-images.githubusercontent.com/48318337/54021277-c5819280-41b5-11e9-95f7-114c94c19a7d.png)

and below URL tells what is expected from application while making query to get custom claims in which Type of Query 

![image](https://user-images.githubusercontent.com/48318337/54021322-e0ec9d80-41b5-11e9-9f57-aa0c9bfc4b39.png)

Where in REF URL : 
![image](https://user-images.githubusercontent.com/48318337/54021360-ffeb2f80-41b5-11e9-8f9f-1355601ae8d4.png)

Tells us while requesting ID Token / Access Token from ADFS, ADFS expect below inputs : 


Parameter | Value
-- | --
response_type | "code"
resource | RP ID (Identifier) of Web API in application group
client_id | client Id of the native application in the application group
redirect_uri | Redirect URI of native application in application group

Where Resource is missing from Loginapp..

https://github.com/fydrah/loginapp/blob/master/main.go

Where from Line 67 to 102.. adding New line to accept Resource as optional Value would be sufficient from config.


`func (s *Server) OAuth2Config(scopes []string) *oauth2.Config {
	return &oauth2.Config{
		ClientID:     s.config.OIDC.Client.ID,
		ClientSecret: s.config.OIDC.Client.Secret,
		RedirectURL:  s.config.OIDC.Client.RedirectURL,
		Endpoint:     s.provider.Endpoint(),
		Scopes:       scopes,
	}
}

// PrepareCallbackURL build and return
// an authCodeURL based on scopes provided
func (s *Server) PrepareCallbackURL() string {
	// Prepare scopes
	var (
		scopes      []string
		authCodeURL string
	)

	scopes = append(scopes, s.config.OIDC.ExtraScopes...)
	// Prepare cross client auth
	// see https://github.com/coreos/dex/blob/master/Documentation/custom-scopes-claims-clients.md
	for _, crossClient := range s.config.OIDC.CrossClients {
		scopes = append(scopes, "audience:server:client_id:"+crossClient)
	}

	scopes = append(scopes, "openid", "profile", "email", "groups")
	if *s.config.OIDC.OfflineAsScope {
		scopes = append(scopes, "offline_access")
		authCodeURL = s.OAuth2Config(scopes).AuthCodeURL(s.config.Name)
	} else {
		authCodeURL = s.OAuth2Config(scopes).AuthCodeURL(s.config.Name)
	}
	logger.Debugf("Request token with the following scopes: %v", scopes)
	return authCodeURL
}
`

**Comment by fydrah** [comment]
@maxh8086 could you try this branch ?
```
docker pull quay.io/fydrah/loginapp:extra-auth-codes-opts
```
You can add the following configuration:
```
extra_auth_code_opts:
  resource: XXXXX
```

**Comment by maxh8086** [comment]
Hello,

i am trying to configure your given image with below configmap in kubernetes, let me know whats wrong with extra_auth_code_opts placement and option that i specify is correct:

  config.yaml: |
    debug: true
    client_id: "4a3f41ff-2fx7-4ec0-b010-b5ax2bf23fx9"
    client_secret: ykRdmYc38h861W9knaHLVeuTlIDaDrB1dHJwAwZP
    issuer_url: "https://adfssrv.domain.com/adfs"
    issuer_root_ca: "/etc/ssl/ca.pem"
    redirect_url: 'https://loginapp.domain.com/callback'
    listen: "https://0.0.0.0:5555"
    tls_cert: "/etc/loginapp/tls/tls.crt"
    tls_key: "/etc/loginapp/tls/tls.key"
    disable_choices: true
    # extra_scopes: "allatclaims"
    app_name: "Kubernetes oAuth2"
    extra_auth_code_opts:
      resource: "https://cluster1.domain.com:6443"
    web_output:
      main_client_id: "4a3f41ff-2fx7-4ec0-b010-b5ax2bf23fx9"
      main_username_claim: "email"

**Comment by maxh8086** [comment]
Correction in representation 
```
extra_auth_code_opts:
   resource: "https://cluster1.domain.com:6443"
web_output:
   main_client_id: "4a3f41ff-2fx7-4ec0-b010-b5ax2bf23fx9"
   main_username_claim: "email"
```

**Comment by maxh8086** [comment]
Till now i tried to make it working without "extra_auth_code_opts:" option but new image is not working.. please let me know if there is something i need to change here..

kubectl logs loginapp-798f995994-vf62c -n auth-system
NAME:
   loginapp - A new cli application

USAGE:
   Web application for Kubernetes CLI configuration with OIDC

COMMANDS:
     serve    Run loginapp application
     help, h  Shows a list of commands or help for one command

OPTIONS:
   --help, -h     show help
   --version, -v  print the version


Present Config file 

```

config.yaml: |
   debug: true
   client_id: "4a3f41ff-2fx7-4ec0-b010-b5ax2bf23fx9"
   client_secret: ykRdmYc38h861W9knaHLVeuTlIDaDrB1dHJwAwZP
   issuer_url: "https://adfssrv.domain.com/adfs"
   issuer_root_ca: "/etc/ssl/ca.pem"
   redirect_url: 'https://loginapp.domain.com/callback'
   listen: "https://0.0.0.0:5555"
   tls_cert: "/etc/loginapp/tls/tls.crt"
   tls_key: "/etc/loginapp/tls/tls.key"
   disable_choices: true
   # extra_scopes: "allatclaims"
   app_name: "Kubernetes oAuth2"
   # extra_auth_code_opts:
   #   resource: "https://cluster1.domain.com:6443"
   web_output:
   main_client_id: "4a3f41ff-2fx7-4ec0-b010-b5ax2bf23fx9"
   main_username_claim: "email"


```

**Comment by fydrah** [comment]
Ho, it seems you use an old configuration format :/

Check [README.md](https://github.com/fydrah/loginapp/tree/extra-auth-codes-opts#configuration) and adapt your configuration.

Also, if you run loginapp in kubernetes, use this as command in your manifest:
```
command:
- "/loginapp"
- "serve"
- "/app/config.yaml"
```

You can also check [this helm chart](https://github.com/ObjectifLibre/kube-oidc) for you deployments, I just updated it a fex days ago.

**Comment by fydrah** [comment]
> Correction in representation
> 
> ```
> extra_auth_code_opts:
>    resource: "https://cluster1.domain.com:6443"
> web_output:
>    main_client_id: "4a3f41ff-2fx7-4ec0-b010-b5ax2bf23fx9"
>    main_username_claim: "email"
> ```

Warning, in the new configuration format, `extra_auth_code_opts` is a key in `oidc` section:
```
oidc:
  extra_auth_code_opts:
    resource: "https://cluster1.domain.com:6443"
```

**Comment by maxh8086** [comment]
Thanks for quick support, Loginapp is working perfect with given modifications. liked the new UI Enhancement and response time.

**Comment by maxh8086** [comment]
one more request.. can you also make some changes to generate a downloadable .kube/config which could be useful in many ways..

like: shell\kubectl, powershell\kubectl as well as for dashboard.

**Comment by fydrah** [comment]
Kubeconfig enhancement is also asked in #13, I will include this in future update.

Currently, do you need extra auth options fields for refresh token (= does "resource" has to be included in kubeconfig as auth-provider-arg ) ?

**Comment by maxh8086** [comment]
This is what it look for Refresh token, Answer was there in same URL :p 


Parameter | Value
-- | --
grant_type | "refresh_token"
resource | RP ID (Identifier) of Web API in application group
client_id | client Id of the native application in the application group
refresh_token | the refresh token issued by AD FS in response to the initial token request

Just an update : 

This What we are getting : 

```
kubectl config set-credentials user@domain.com \
    --auth-provider oidc \
    --auth-provider-arg idp-issuer-url= \
    --auth-provider-arg client-id= \
    --auth-provider-arg id-token= \
    --auth-provider-arg client-secret= \
    --auth-provider-arg refresh-token=
```

This is what could be expected result as 1 time human intervention just to access cluster and reduce rest of the efforts. :

```
kubectl config set-cluster $ClusterName --certificate-authority=API-CA.pem --embed-certs=true --server=https://API-URL:6443

kubectl config set-context test-cluster-context --cluster=$ClusterName --user=user@domain.com 

kubectl config set-credentials user@domain.com \
   --auth-provider=oidc \
   --auth-provider-arg=idp-issuer-url= \
   --auth-provider-arg=client-id= \
   --auth-provider-arg=client-secret= \
   --auth-provider-arg=refresh-token= \
   --auth-provider-arg=idp-certificate-authority=/tmp/API-CA.pem \
   --auth-provider-arg=id-token=

kubectl config use-context test-cluster-context
```

**Note : Not sure if "--auth-provider-arg=idp-certificate-authority=/tmp/API-CA.pem \" this line is required if certificate is already provided under "kubectl config set-cluster" line.**

**Comment by maxh8086** [comment]
i am not sure if kubectl would go an and request for refresh token, dashboard may do.. 
considering kubectl as Native application here and dashboard as web application.

**Comment by fydrah** [comment]
Hi @maxh8086 ,

You can try this new image:
```
docker pull quay.io/fydrah/loginapp:2.6.0-rc1
```

This probably be the next release.

It does not include `kubectl config set-cluster ....` enhancement, as I said I will work on it soon.

**Comment by maxh8086** [comment]
Have deployed new image, will check on monday if token does auto refresh by kubectl.

have gone through docuement where it says, kubectl 1.3 would be able to refresh token.

**Comment by maxh8086** [comment]
Hello Flavin,

Sorry for delayed response, i have tested refresh token functionality today. however it seems to be not working as expected with ADFS. ADFS by default provide 1 hour validity to issued token, however after validating post 1 hr but before 2 hrs Token was expired and was not refreshed automatically.

Need to understand kubectl behavior to get new refresh token from ADFS automatically, not yet clear on that part.

**Comment by fydrah** [comment]
As said [here](https://docs.microsoft.com/fr-fr/windows-server/identity/ad-fs/overview/ad-fs-scenarios-for-developers#single-sign-on-behavior), you should check SSO_period for refresh tokens.

Also:
```
Providers that don’t return an id_token as part of their refresh token response aren’t supported
```
(https://kubernetes.io/docs/reference/access-authn-authz/authentication/#option-1-oidc-authenticator)
I hope adfs does, but you should check this too.

**Comment by fydrah** [comment]
Closing this issue since feature has been released into version 2.6.0 .
Re-open if needed.

I forgot to answer to you:
> i am not sure if kubectl would go an and request for refresh token, dashboard may do.. considering kubectl as Native application here and dashboard as web application.

If loginapp do this, it must store refresh token. This is not the purpose of this application, it is just a frontend to request tokens to dex and give them back to users with a friendly web page, not to handle authentication lifecycle.

Regards,
